core/commands/store/contracts/contracts.go

```diff
@@ -174,47 +189,126 @@ This command get contracts list based on role from the local node data store.`,
+			result = append(result, c)
 		}
-		return cmds.EmitOnce(res, &nodepb.Contracts{Contracts: data})
+		return cmds.EmitOnce(res, &nodepb.Contracts{Contracts: result})
 	},
 	Type: nodepb.Contracts{},
 }
+
+func getKey(role string) string {
+	var k string
+	if role == nodepb.ContractStat_HOST.String() {
+		k = hostContractsKey
+	} else if role == nodepb.ContractStat_RENTER.String() {
+		k = renterContractsKey
+	} else {
+		return "reserved"
+	}
+	return k
+}
+
+func Save(d datastore.Datastore, cs []*nodepb.Contracts_Contract, role string) error {
+	return ds.Save(d, getKey(role), &contractspb.Contracts{
+		Contracts: cs,
+	})
+}
+
+func ListContracts(d datastore.Datastore, role string) ([]*nodepb.Contracts_Contract, error) {
+	cs := &contractspb.Contracts{}
+	err := ds.Get(d, getKey(role), cs)
+	if err != nil && err != datastore.ErrNotFound {
+		return nil, err
+	}
+	return cs.Contracts, nil
+}
+
+func SyncContracts(ctx context.Context, n *core.IpfsNode, role string) error {
+	cfg, err := n.Repo.Config()
+	if err != nil {
+		return err
+	}
+	pk, err := n.Identity.ExtractPublicKey()
+	if err != nil {
+		return err
+	}
+	pkBytes, err := pk.Raw()
+	if err != nil {
+		return err
+	}
+	results := make([]*nodepb.Contracts_Contract, 0)
+	cs, err := ds.ListShardsContracts(n.Repo.Datastore(), n.Identity.Pretty(), role)
+	if err != nil {
+		return err
+	}
+	err = grpc.EscrowClient(cfg.Services.EscrowDomain).WithContext(ctx,
+		func(ctx context.Context,
+			client escrowpb.EscrowServiceClient) error {
+			for _, c := range cs {
+				ec, err := escrow.UnmarshalEscrowContract(c.SignedEscrowContract)
+				if err != nil {
+					log.Error("unmarshal escrow contract error:", err)
+					continue
+				}
+				in := &escrowpb.SignedContractID{
+					Data: &escrowpb.ContractID{
+						ContractId: ec.Contract.ContractId,
+						Address:    pkBytes,
+					},
+				}
+				sign, err := crypto.Sign(n.PrivateKey, in.Data)
+				if err != nil {
+					log.Error("sign contractID error:", err)
+					continue
+				}
+				in.Signature = sign
+				s, err := client.GetPayOutStatus(ctx, in)
+				if err != nil {
+					log.Error("get payout status error:", err)
+					continue
+				}
+				results = append(results, &nodepb.Contracts_Contract{
+					ContractId:              c.GuardContract.ContractId,
+					HostId:                  c.GuardContract.HostPid,
+					RenterId:                c.GuardContract.RenterPid,
+					Status:                  c.GuardContract.State,
+					StartTime:               c.GuardContract.RentStart,
+					EndTime:                 c.GuardContract.RentEnd,
+					NextEscrowTime:          s.Status.NextPayoutTime,
+					CompensationPaid:        s.Status.PaidAmount,
+					CompensationOutstanding: s.Status.Amount - s.Status.PaidAmount,
+					UnitPrice:               c.GuardContract.Price,
+					ShardSize:               c.GuardContract.ShardFileSize,
+					ShardHash:               c.GuardContract.ShardHash,
+					FileHash:                c.GuardContract.FileHash,
+				})
```

Comment:
this could go into go-btfs-common OR part of core/escrow

Reply:
okay, will refactor this after merged.
i'll raise a pr for go-btfs-common.

Reply:
okay